### PR TITLE
cephfs: refactor cephfs core functions

### PR DESCRIPTION
**Pull request by Madhu-1**
This commits refactors the cephfs core functions with interfaces.

Signed-off-by: Madhu Rajanna <madhupr007@gmail.com>

### CODE REVIEW

File: internal/cephfs/core/clone.go [hunk 1]
```diff
@@ -47,28 +46,28 @@ const (
...
 	ctx context.Context,
-	volID, cloneID fsutil.VolumeID,
-	volOpt,
-	parentvolOpt *VolumeOptions) error {
-	snapshotID := cloneID
-	err := parentvolOpt.CreateSnapshot(ctx, snapshotID, volID)
+	parentvolOpt *SubVolume) error {
+	snapshotID := s.VolID
+	snapClinet := NewSnapshot(s.conn, snapshotID, parentvolOpt)
+	err := snapClinet.CreateSnapshot(ctx)
```
Comment (nixpanic) [inline internal/cephfs/core/clone.go]: `snapClinet` ?

File: internal/cephfs/core/volume.go [hunk 2]
```diff
@@ -53,20 +53,74 @@ type Subvolume struct {
...
+	CreateCloneFromSubvolume(ctx context.Context, parentvolOpt *SubVolume) error
+	// GetCloneState returns the clone state of the subvolume.
+	GetCloneState(ctx context.Context) (cephFSCloneState, error)
+	// CreateCloneFromSnapshot creates a clone from the subvolume snapshot.
+	CreateCloneFromSnapshot(ctx context.Context, snap Snapshot) error
+	// CleanupSnapshotFromSubvolume removes the snapshot from the subvolume.
+	CleanupSnapshotFromSubvolume(ctx context.Context, parentVol *SubVolume) error
+}
+
+type subVolumeClient struct {
```
Comment (nixpanic) [inline internal/cephfs/core/volume.go:84]: Adding comments for new types, and their attributes would be much appreciated!

File: internal/cephfs/store/volumeoptions.go [hunk 3]
```diff
@@ -25,42 +25,39 @@ import (
...
 	Mounter            string `json:"mounter"`
 	ProvisionVolume    bool   `json:"provisionVolume"`
 	KernelMountOptions string `json:"kernelMountOptions"`
 	FuseMountOptions   string `json:"fuseMountOptions"`
-	SubvolumeGroup     string
-	Features           []string
 
 	// conn is a connection to the Ceph cluster obtained from a ConnPool
-	conn *util.ClusterConnection
+	Conn *util.ClusterConnection
```
Comment (nixpanic) [inline internal/cephfs/store/volumeoptions.go]: This should really be internal... what is the reason to export it outside of the package? If it is only temporary, add a `FIXME` comment.

File: internal/cephfs/store/fsjournal.go [hunk 4]
```diff
@@ -14,13 +14,14 @@ See the License for the specific language governing permissions and
 limitations under the License.
 */
 
-package core
+package store
```
Comment (nixpanic) [inline internal/cephfs/store/fsjournal.go:17]: moving files to a new package should probably be a separate commit?

File: internal/cephfs/store/volumeoptions.go [hunk 5]
```diff
@@ -579,23 +579,34 @@ func NewSnapshotOptionsFromID(
...
 	if err != nil {
 		return &volOptions, nil, &sid, err
 	}
 
 	return &volOptions, &info, &sid, nil
 }
 
-func GenSnapFromOptions(ctx context.Context, req *csi.CreateSnapshotRequest) (snap *CephfsSnapshot, err error) {
-	cephfsSnap := &CephfsSnapshot{}
+type SnapshotOption struct {
```
Comment (nixpanic) [inline internal/cephfs/store/volumeoptions.go:605]: comments for the struct and members?

**Review by nixpanic — CHANGES_REQUESTED** [review]
Quite a large PR, but most things look trivial. Reviewing and adding unit-tests would be easier when it is split in a few commits.

File: internal/cephfs/core/volume.go (hunk 2 above)
Comment (Madhu-1) [reply internal/cephfs/core/volume.go:84]: sure added it.

File: internal/cephfs/store/volumeoptions.go (hunk 3 above)
Comment (Madhu-1) [reply internal/cephfs/store/volumeoptions.go]: i was not able to use the unexported member in other files. added a helper function to get the connection made it as private again.

File: internal/cephfs/store/fsjournal.go (hunk 4 above)
Comment (Madhu-1) [reply internal/cephfs/store/fsjournal.go:17]: as there is inter connection between a lot of things I kept it as a single commit.

**Comment by Madhu-1** [comment]
> Quite a large PR, but most things look trivial. Reviewing and adding unit-tests would be easier when it is split in a few commits.

@nixpanic I thought to make multiple commits. as most of the things are interconnected I kept it as a single commit for easy rebase and cherry-picking if required. Hope should be fine.

**Review by nixpanic — APPROVED** [review]

**Review by yati1998 — COMMENTED** [review]
LGTM but will leave approval rights for others.

**Comment by Madhu-1** [comment]
@ceph/ceph-csi-contributors anyone else who can help with review/approval?

**Review by Rakshith-R — APPROVED** [review]

**Comment by Rakshith-R** [comment]
@mergifyio rebase

**Comment by Rakshith-R** [comment]
@Mergifyio rebase

**Comment by Madhu-1** [comment]
@mergifyio rebase

**Comment by Madhu-1** [comment]
> /retest all

CI seems stuck restarting all tests
